### PR TITLE
Support execution by partition in addition to by bucket for Hive connector

### LINKED ISSUE #930: Grouped execution on partition keys
Today, grouped ("bucketed") execution is supported on bucketed tables.
Add support for grouped execution for partitioned unbucketed tables.

For example, joining partitioned tables on partition key can be handled the same way as joining bucketed tables on bucketing key.

Ref https://github.com/prestodb/presto/issues/11876

### PR DESCRIPTION
**Pull request by jacobbaskin**
## Description

Just as it can be helpful to do bucketed execution, execution by partition can improve performance in a number of situations, especially aggregation by partition, and enables colocated joins where the partition key is part of the join key. In particular, with fault-tolerant execution, this means that it is possible to join two tables by partition without persisting either side of the join to the exchange.

The main limitation to the feature as implemented in this PR is that it is all-or-nothing with respect to partition columns. That is, if a join uses one partitioned column but the table is partitioned on two columns, execution by partition cannot be used. Similarly, if a table is partitioned and bucketed, both the partition key and bucketing key must be used in a join.

> Is this change a fix, improvement, new feature, refactoring, or other?

New feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Hive connector only

> How would you describe this change to a non-technical end user or system administrator?

If you have a large table with many partitions, this makes it easier to query the table or join it with other tables in an efficient way.

## Related issues, pull requests, and links
* Fixes #930 (which references grouped execution; this feature is still useful without grouped execution, however.)

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Hive Connector
* Partitioned execution support (similar to bucketed execution, but for partitioned tables)
```

### CODE REVIEW

**Comment by findepi** [comment]
cc @electrum @dain

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:125]: this is someting we might want to enable by default if benchmarks results are good. cc @raunaqmorarka @przemekak

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionedBucketFunction.java
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionedBucketFunction.java:36]: where this one went?

Comment (jacobbaskin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionedBucketFunction.java:36]: Sorry it was renamed but I screwed up the commit history, should be clearer now

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java
Comment (jacobbaskin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:125]: curious how you'd scale this by default...if you have too few partitions the reduction in parallelism will outweigh the improved locality.

**Comment by jacobbaskin** [comment]
@sopel39 this should be ready for review now; it passes tests locally and new tests demonstrate that execution by partition works for both bucketed and non-bucketed tables. Let me know if there is anything else I should add.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java:160]: I think this might be redundant to `optimizer.table-scan-node-partitioning-min-bucket-to-task-ratio`. Could you verify?

Comment (jacobbaskin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java]: Good point, changed to a boolean property.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveBucketFunction.java
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveBucketFunction.java:13]: Commit message is too long

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:127]: It should be on by default. Otherwise Noone will use it

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveNodePartitioningProvider.java
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveNodePartitioningProvider.java:121]: Don't use `final` on local variables. Declare as the interface type `Map` rather than the implementation `ImmutableMap`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java:556]: "else" here is redundant

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java:960]: This seems to duplicate `toPartitionValues`. Can we combine them?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveBucketFunction.java
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveBucketFunction.java:44]: Private helper methods should go at the bottom

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveBucketFunction.java
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveBucketFunction.java:41]: This is effectively making `HiveBucketFunction` a union type which is more complicated and harder to understand. Instead, let's create a separate implementation for partitions.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveNodePartitioningProvider.java
Comment (jacobbaskin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveNodePartitioningProvider.java]: Done.

**Review by electrum — CHANGES_REQUESTED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
Comment (jacobbaskin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java:960]: Changed to implement `erasePartitionColumnNames` in terms of `toPartitionValues`.

File: testing/trino-tests/src/test/resources/sql/presto/tpcds/hive/partitioned/q01.plan.txt
Comment (raunaqmorarka) [inline testing/trino-tests/src/test/resources/sql/presto/tpcds/hive/partitioned/q01.plan.txt:17]: A lot of TPC plans have changed from
```
join (INNER, REPLICATED):
    scan partitioned_fact_table
    local exchange (GATHER, SINGLE, [])
         remote exchange (REPLICATE, BROADCAST, [])
             scan date_dim
```             
to
```
join (INNER, PARTITIONED):
    scan partitioned_fact_table
    local exchange (GATHER, SINGLE, [])
         remote exchange (REPARTITION, local:HivePartitioningHandle{buckets=1, hiveTypes=[], partitions=2004}, ["d_date_sk"])
             scan date_dim
```
That might be a worse plan because usually we want to broadcast small build side rather than repartition them.

Comment (jacobbaskin) [reply testing/trino-tests/src/test/resources/sql/presto/tpcds/hive/partitioned/q01.plan.txt:17]: What would you suggest here? The same thing would happen in this case if the probe table was bucketed, so this seems to me like a more general planner change. I was hoping to keep this PR in the Hive connector rather than the planner as much as possible. Maybe we can profile and see how urgent this problem is. Another option is disabling this feature by default pending planner improvements.

Comment (jacobbaskin) [reply testing/trino-tests/src/test/resources/sql/presto/tpcds/hive/partitioned/q01.plan.txt:17]: Following up on this...anything I can do? I have rebased the PR against current master.

Comment (raunaqmorarka) [reply testing/trino-tests/src/test/resources/sql/presto/tpcds/hive/partitioned/q01.plan.txt:17]: Ideally we want features to ship as enabled by default so that all users benefit from it and it gets sufficient validation in  a variety of real world workloads.
If other maintainers agree, we could choose to keep this disabled by default.
However, this still requires proving performance benefits of this optimization and having green CI with this feature turned on so that we know there aren't surprises waiting for us when we try to enable it by default later.
E.g. It looks like right now there are product test failures with transactional tables, this needs to be resolved.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionedBucketFunction.java
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionedBucketFunction.java:58]: I saw division by zero failures here when executing benchmarks
```
2023-03-06T08:54:46.3906080Z Caused by: io.trino.jdbc.$internal.client.FailureInfo$FailureException: / by zero
2023-03-06T08:54:46.3906613Z 	at io.trino.plugin.hive.util.HiveBucketing.getBucketNumber(HiveBucketing.java:172)
2023-03-06T08:54:46.3907179Z 	at io.trino.plugin.hive.util.HiveBucketing.getHiveBucket(HiveBucketing.java:125)
2023-03-06T08:54:46.3907740Z 	at io.trino.plugin.hive.HiveBucketFunction.getBucket(HiveBucketFunction.java:47)
2023-03-06T08:54:46.3908373Z 	at io.trino.plugin.hive.HivePartitionedBucketFunction.lambda$getBucket$0(HivePartitionedBucketFunction.java:58)
2023-03-06T08:54:46.3908921Z 	at java.base/java.util.Optional.map(Optional.java:260)
2023-03-06T08:54:46.3910479Z 	at io.trino.plugin.hive.HivePartitionedBucketFunction.getBucket(HivePartitionedBucketFunction.java:58)
2023-03-06T08:54:46.3913060Z 	at io.trino.operator.BucketPartitionFunction.getPartition(BucketPartitionFunction.java:46)
2023-03-06T08:54:46.3913974Z 	at io.trino.operator.output.PagePartitioner.lambda$partitionPositions$2(PagePartitioner.java:244)
2023-03-06T08:54:46.3914629Z 	at io.trino.operator.output.PagePartitioner.partitionNotNullPositions(PagePartitioner.java:387)
2023-03-06T08:54:46.3916236Z 	at io.trino.operator.output.PagePartitioner.partitionGeneric(PagePartitioner.java:353)
2023-03-06T08:54:46.3916862Z 	at io.trino.operator.output.PagePartitioner.partitionPositions(PagePartitioner.java:244)
2023-03-06T08:54:46.3917505Z 	at io.trino.operator.output.PagePartitioner.partitionPageByColumn(PagePartitioner.java:203)
2023-03-06T08:54:46.3918122Z 	at io.trino.operator.output.PagePartitioner.partitionPage(PagePartitioner.java:150)
2023-03-06T08:54:46.3918892Z 	at io.trino.operator.output.PartitionedOutputOperator.addInput(PartitionedOutputOperator.java:312)
```

**Comment by mosabua** [comment]
@raunaqmorarka and @jacobbaskin .. what is the status of this work currently. I am not sure who is waiting for whom. I think it would be good to move this forward to a merge.

**Comment by jacobbaskin** [comment]
Here is the status as far as I know:
- Some tests broke when I last rebased; I can't reproduce locally. I gave
it a try for a while but ended up working on other things.
- Also, I haven't looked into Raunaq's division by zero failure above,
which would obviously be a prerequisite to merging.
- Also, someone has to decide if the plan changes above are positive or
negative (i.e., more repartition joins and fewer broadcast joins)

The first two are on me; I can spend some dedicated time on these next
week. I'm still very excited to get this merged. If anyone can help on
repro-ing the CI failures I'd appreciate it but I can also look into these
on my own. I don't know who would be the right decider on the third thing.

**Comment by mosabua** [comment]
@martint @arhimondr and myself had a good chat with @jacobbaskin and we have a plan to move this forward.

**Comment by jacobbaskin** [comment]
@martint @arhimondr so I rebased and disabled partition execution by default; the build is now green. Should we merge or anything else you wanted to check out first?

**Comment by arhimondr** [comment]
@jacobbaskin I may not be able to get to this for the next couple of weeks. Please don't wait for me.

**Comment by devozerov** [comment]
Hi @jacobbaskin. I have a couple of questions for this PR if you don't mind.
1. If we are doing reads, the bucket is determined as (a) extract the partition names from split/row, (b) get partition position from the sorted list of partition names. What is the advantage of this approach over simply doing `hash(partitionNames) % maxPartitionCount` ? The advantage of the latter is that we do not need to rely on partition names completely. AFAIU this is already do in the write-related bucket function, but there is a comment that this strategy is not good enough. What is the motivation for this statement?
2. In `getCommonPartitioningHandle`, the partitioning is not used if two tables have different partitions. What is the motivation for this? Consider the query `SELECT t1 JOIN t2 ON t1.part_col1 = t2.part_col2 WHERE t1.part_col1 in (1, 2) AND t2.part_col2 in (2, 3)`. If both columns have the same types for their partitioning columns, the partitioned execution is technically possible (just route splits from the same partitions to the same nodes). However, the current implementation seems to disallow this. Why is this the case? 

I am asking because we implemented the same feature in our private fork, and we see that is it somewhat common when several fact table with different effective partitions are merged/"unioned" and we always allow for co-partitioned execution in this case because it reduces exchange burden.

**Comment by jacobbaskin** [comment]
@devozerov:
For your first question, I had two motivations.

One is the birthday problem: if the number of partitions and nodes is of the same order of magnitude, the chance is very high that you will end up with a hash collision, and one node having 2x the work (and, for colocated joins, 2x the memory requirement) of the others. For instance, with 10 partitions and 20 nodes, there is a >93% chance of a collision. Even with 5 partitions and 20 nodes, there is a 41% chance of a collision!

The second is fault-tolerant execution, which removes the necessity for chunks of work to be divided into nodes: you could have more tasks than nodes and just not schedule them all at once. In this case, it is extra helpful to know what the "natural" bucketing of the data is to enable efficient task scheduling.

For your second question, I don't think there's anything stopping us from doing this, I just didn't implement it. I was worried about introducing correctness bugs, and my understanding of the query planner is that, in most cases, partitions 1 and 3 will be dropped before `getCommonPartitioningHandle` is called (see my comment on line 3096). Would love to see you upstream an extension to this once the PR gets merged (hopefully soon!)

**Comment by raunaqmorarka** [comment]
I ran benchmarks with this feature enabled again
The main concerning thing is that 25 TPCDS queries failed 
[Partitioned execution sf1000 partitioned parquet.pdf](https://github.com/trinodb/trino/files/11512957/Partitioned.execution.sf1000.partitioned.parquet.pdf)

Logs
[Partitioned execution.log](https://github.com/trinodb/trino/files/11512971/Partitioned.execution.log)

There are failures due to
```
2023-05-18T18:44:09.1147553Z Caused by: io.trino.jdbc.$internal.client.FailureInfo$FailureException: / by zero
2023-05-18T18:44:09.1148237Z 	at io.trino.plugin.hive.util.HiveBucketing.getBucketNumber(HiveBucketing.java:172)
2023-05-18T18:44:09.1148972Z 	at io.trino.plugin.hive.util.HiveBucketing.getHiveBucket(HiveBucketing.java:125)
2023-05-18T18:44:09.1149861Z 	at io.trino.plugin.hive.HiveBucketFunction.getBucket(HiveBucketFunction.java:47)
2023-05-18T18:44:09.1150712Z 	at io.trino.plugin.hive.HivePartitionedBucketFunction.lambda$getBucket$0(HivePartitionedBucketFunction.java:58)
2023-05-18T18:44:09.1151426Z 	at java.base/java.util.Optional.map(Optional.java:260)
2023-05-18T18:44:09.1152163Z 	at io.trino.plugin.hive.HivePartitionedBucketFunction.getBucket(HivePartitionedBucketFunction.java:58)
2023-05-18T18:44:09.1153575Z 	at io.trino.operator.BucketPartitionFunction.getPartition(BucketPartitionFunction.java:46)
2023-05-18T18:44:09.1154408Z 	at io.trino.operator.output.PagePartitioner.lambda$partitionPositions$2(PagePartitioner.java:252)
2023-05-18T18:44:09.1155277Z 	at io.trino.operator.output.PagePartitioner.partitionNotNullPositions(PagePartitioner.java:395)
2023-05-18T18:44:09.1156134Z 	at io.trino.operator.output.PagePartitioner.partitionGeneric(PagePartitioner.java:362)
2023-05-18T18:44:09.1157170Z 	at io.trino.operator.output.PagePartitioner.partitionPositions(PagePartitioner.java:252)
2023-05-18T18:44:09.1158049Z 	at io.trino.operator.output.PagePartitioner.partitionPageByColumn(PagePartitioner.java:211)
2023-05-18T18:44:09.1159022Z 	at io.trino.operator.output.PagePartitioner.partitionPage(PagePartitioner.java:158)
2023-05-18T18:44:09.1160069Z 	at io.trino.operator.output.PartitionedOutputOperator.addInput(PartitionedOutputOperator.java:335)
```
There are also incorrect results
```
2023-05-18T20:01:15.8038761Z 20:01:15.814 INFO  [pool-3-thread-1] i.t.benchto.driver.utils.QueryUtils - Comparing result with results/tpcds/sf1000/q75.csv
2023-05-18T20:01:33.0444489Z 20:01:33.054 ERROR [pool-3-thread-1] i.t.b.d.e.BenchmarkExecutionDriver - Query Execution failed for benchmark presto/tpcds query q75
2023-05-18T20:01:33.0446592Z io.trino.benchto.driver.execution.ResultComparisonException: Incorrect result at row 1, expected 2001,2002,3001001,1,9,173,910432,310842,-599590,-25754403.98, got 2001,2002,3001001,1,9,173,273518,93391,-180127,-8452353.74
```

**Comment by jacobbaskin** [comment]
@raunaqmorarka sorry for the delay. I tested locally and I believed I fixed the bug that you were seeing with TPCDS data, which was a really embarassing one I should have caught earlier :-) I'll wait for the build to be green and assuming that works would be grateful if you reran benchmarks. Thanks!

**Comment by raunaqmorarka** [comment]
I re-ran benchmarks after latest fixes. I no longer see query failures or result correctness issues.
There are some good gains in TPCH. 
In TPCDS there are significant wall time regressions (probably due to not enough distribution of work across nodes). There is a large CPU time regression in q72 which is concerning and worth investigating, the rest is not too bad if we keep this flag disabled by default.
Screenshot 2023-06-16 at 2 43 20 PM

[Partitioned execution sf1k parquet.pdf](https://github.com/trinodb/trino/files/11768658/Partitioned.execution.sf1k.parquet.pdf)

**Comment by sopel39** [comment]
10% for TPCH. Nice!. I wonder if we can improve CBO rule so that there are not regressions or gains in TPC-DS

**Comment by mosabua** [comment]
@jacobbaskin could you resolve the conflicts?

@sopel39 @raunaqmorarka @martint .. is this good enough to go in soon? It is an optional feature for experts with huge performance gains in some cases and imho worth having as is. We can always refine more and look towards switching it to on by default later.

**Comment by mosabua** [comment]
Wdyt @raunaqmorarka @sopel39 @martint  .. can we merge that in now with the 2 week window to 423 in August

**Comment by sopel39** [comment]
> Note that DFs are not supported in partitioned table scan.

BTW: I don’t think it’s a blocker for https://github.com/trinodb/trino/pull/13319, because DFs will work in general, but they won’t be lazy

**Comment by mosabua** [comment]
:wave: @jacobbaskin @raunaqmorarka @sopel39 @martint  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by mosabua** [comment]
@jacobbaskin could you check out https://github.com/trinodb/trino/pull/23432 and the upcoming 468 release and see if that work from @dain helps .. or alternative check in with us to discuss how we can drive the improvements from your work here into Trino with our ongoing performance work

cc @raunaqmorarka @sopel39 @martint

**Comment by jacobbaskin** [comment]
No, this doesn’t help us. On the other hand, I am relatively more skeptical
of the need for our initial feature compared to two years ago. It turns out
exchange is a smaller cost than I thought it would be, and the chance of
getting skewed workloads is larger.

**Pull request closed without merging**
